Database metric semantic conventions (#1076)
open-telemetry/opentelemetry-specification: specification/metrics/semantic_conventions/database.md

# Semantic Conventions for Database Metrics

> 

Reviewer: ```suggestion

<!-- Re-generate TOC with `markdown-toc --no-first-h1 -i` -->

<!-- toc -->

Run the tool to have the TOC generated here

<!-- tocstop -->

```

  "labels": {
    "db.operation": "SELECT",
    "db.table": "user_table",
    "db.name": "user_db",
    "db.system": "postgresql",
    "db.connection_string": "postgresql://postgres-server:5432/user_db",
    "db.user": "",
    "net.peer.ip": "192.0.10.2",
    "net.peer.port": 5432,
    "net.peer.name": "postgres-server"
  }
}
```

#### MySQL SELECT Query

For a client executing a query like this:

```SQL
SELECT * FROM orders WHERE order_id = 301;
```

while connected to a MySQL database named "ShopDb" running on host
`shopdb.example.com`, the following instrument should result:

```json
{
  "name": "db.client.duration",
  "labels": {
    "db.operation": "SELECT",
    "db.table": "orders",
    "db.name": "ShopDb",
    "db.system": "mysql",
    "db.connection_string": "Server=shopdb.example.com;Database=ShopDb;Uid=billing_user;TableCache=true;UseCompression=True;MinimumPoolSize=10;MaximumPoolSize=50;",
    "db.user": "billing_user",
    "net.peer.name": "shopdb.example.com",
    "net.peer.ip": "192.0.2.12",
    "net.peer.port": "3306",
    "net.transport": "IP.TCP"
  }
}

```

#### Redis HMSET

For a client executing a Redis HMSET command like this:

```redis

> HMSET myhash field1 'Hello' field2 'World

Reviewer: ```suggestion
HMSET myhash field1 'Hello' field2 'World'
```

Author: Fixed in deea045

```

while connecting to a Redis instance over a Unix socket, the following instrument
should result:

```json
{
  "name": "db.client.duration",
  "labels": {
    "db.operation": "HMSET",
    "db.table": "myhash",
    "db.system": "redis",
    "db.user": "the_user",
    "net.peer.name": "/tmp/redis.sock",
    "net.transport": "Unix",
    "db.redis.database_index": "15"
  }
}
```

#### MongoDB findAndModify

For a mongo client executing the `findAndModify` command like this:

```javascript
{
  findAndModify: "people",
  query: { name: "Tom", state: "active", rating: { $gt: 10 } },
  sort: { rating: 1 },
  update: { $inc: { score: 1 } }
}
```

against the database named "userdatabase" while connected to a MongoDB available
at `mongodb.example.com`, the following instrument should result:

```json
{
  "name": "db.client.duration",
  "labels": {
    "db.operation": "findAndModify",
    "db.table": "people",
    "db.name": "userdatabase",
    "db.system": "mongodb",
    "db.user": "the_user",
    "net.peer.name": "mongodb.example.com",
    "net.peer.ip": "192.0.2.14",
    "net.peer.port": "27017",
    "net.transport": "IP.TCP"
  }
}
```

## Connection Pooling Metric Instruments

> Otherwise, the following metric instruments SHOULD be collected. They SHOULD

Reviewer 2: What's the "Otherwise" in relation to?

Author: :man_facepalming:

Author: Fixed in deea045

have all [common](#common) labels applied to them.

| Name                      | Instrument    | Units        | Description |
|---------------------------|---------------|--------------|-------------|
| `db.connection_pool.limit` | ValueObserver | {connections} | The total number of database connections available in the connection pool. |
| `db.connection_pool.usage` | ValueObserver | {connections} | The number of database connections _in use_. |

> If the following detailed information is available, instrumentation MAY collect

Reviewer 2: Maybe change "instrumentation MAY collect the following metric instruments" to "the following metric instruments MAY be collected", as the collection wouldn't only apply to instrumentation.

Author: Fixed in deea045

the following metric instruments. They SHOULD have all [common](#common) labels
applied to them.

| Name                      | Instrument | Units         | Description |
|---------------------------|------------|---------------|-------------|
| `db.connections.new`      | Counter   | {connections} | The number of new connections created. |
| `db.connections.taken`    | Counter   | {connections} | The number of connections taken from the connection pool. |

> | `db.connections.returned` | Counter   | {connections} | The number of connections returned to the connection pool. |

Reviewer 2: Is it anticipated that `db.connection_pool.usage` is equal to `db.connections.taken` minus `db.connections.returned`?

If so, what is the benefit in having the additional instruments?

Author: Yes, I expect `db.connection_pool.usage` = `db.connections.taken` - `db .connections.returned`. But the breakdown metrics can give much more detail than the use of the connection pool over time.  
The usage/limit metrics use a LastValue aggregator by default, meaning you get a periodic snapshot of the number of connections currently in use.
The more detailed metrics allow you to see things like connection churn. With them, a user could differentiate between an application making millions of tiny database operations (and ending a harvest period with 4/5 connections in use) and an application with four very long and slow queries in progress (also ending the harvest period with 4 connections in use).

The connection_pool usage instruments are asynchronous -- they're intended to be polled on a set interval. The more detailed metrics are recording all the actions related to the connection pool in real time.  

*...but*
Having said all of that, I don't have specific use cases to back up the inclusion of these extra instruments. 
_But_, they're already being recorded by go-redis, and I can see the same sort of instruments being possible in a number of other frameworks. And if there's a desire to record them, I think we should spec them so we get consistent names and labels across implementations.

Author: I tried a few different ways of explaining this. I ultimately ended up with the pretty simple approach in ec58de. 
I'm certainly open to suggestions about how to clarify what these instruments are all about.

> | `db.connections.reused`   | Counter   | {connections} | The number of connections reused. |

Reviewer 2: What are the criteria for knowing it's an existing connection that's being re-used?

Author: This is a good point.  I had used [this PR](https://github.com/go-redis/redis/pull/1427) as reference, and had assumed that an instrumentation author would know best how to collect these numbers, but now I'm thinking we need definitions of exactly what value needs to be recorded here, or we run the risk of misunderstanding and inconsistency across instrumentation sources.  

I'll give this some thought about how to describe this.

# Semantic Conventions for Database Metrics

<!-- Re-generate TOC with `markdown-toc --no-first-h1 -i` -->

<!-- toc -->

> Run the tool to have the TOC generated here

Reviewer: Oh this was just a placeholder 😄 
You'll actually have to run the tool locally on your working copy to make the magic happen. Our CI here doesn't alter any files that are checked in.

Database operations SHOULD include execution of queries, including DDL, DML,
DCL, and TCL SQL statements (and the corresponding operations in non-SQL
databases), as well as connect operations.

### Labels

In addition to the [common](#common) labels, the following labels SHOULD be
applied to all database call-level metric instruments.

| Label Name       | Type   | Description  | Example  | Required |
|------------------|--------|--------------|----------|----------|
| `db.name`        | string | If no [tech-specific label](#call-level-labels-for-specific-technologies) is defined, this attribute is used to report the name of the database being accessed. For commands that switch the database, this should be set to the target database (even if the command fails). [1] | `customers`<br>`main` | Required if applicable. |
| `db.operation`   | string | The name of the operation being executed, e.g. the [MongoDB command name](https://docs.mongodb.com/manual/reference/command/#database-operations) such as `findAndModify`. [4][5] | `findAndModify`<br>`HMSET`<br>`SELECT`<br>`CONNECT` | Required if applicable. |

> | `db.table`       | string | The name of the primary table, collection, segment, etc... that the operation is acting upon. | `user_table` | Required if applicable. |

Reviewer: I'd be in favor of introducing this in the (more elaborate) tracing spec first, as you proposed on #1104, and then just refer to that definition from here to avoid duplication and potentially going out of sync. Once both kinds of semconvs are auto-generated, it should be fine again, however.

Author: :100: I've created that PR now (here: #1141). After it has some reviews, I'll update this PR like you suggest.

Author: PR #1141 has all the approvals it needs, so I've updated this one to match it. 
Just that little bit of manual work to keep these two PRs has reminded me that we really need the semantic convention table generator to work for metrics.

represeting the statement or operation being executed on the database. It may be
a stored procedure name (without arguments), operation name, etc.

### Call-level labels for specific technologies

| Label Name                | Description  | Example  | Required |
|---------------------------|--------------|----------|----------|
| `db.cassandra.keyspace`   | The name of the keyspace being accessed. To be used instead of the generic `db.name` attribute. | `mykeyspace` | Yes |
| `db.hbase.namespace`      | The [HBase namespace](https://hbase.apache.org/book.html#_namespace) being accessed. To be used instead of the generic `db.name` attribute. | `default` | Yes |
| `db.redis.database_index` | The index of the database being accessed as used in the [`SELECT` command](https://redis.io/commands/select), provided as an integer. To be used instead of the generic `db.name` label. | `0`<br>`1`<br>`15` | Conditional [1] |
| `db.mongodb.collection`   | The collection being accessed within the database stated in `db.name`. | `customers`<br>`products` | Yes |

**[1]:** Required, if other than the default database (`0`).

> ### Examples

Reviewer: Examples are great indeed, thanks!
No need to put that praise into parentheses, @jmacd! 😄

#### PostgreSQL SELECT Query

For a client executing a query like this:

```SQL
SELECT * FROM public.user_table WHERE user_id = 301;
```

while connected to a PostgreSQL database named "user_db" running on host
`postgres-server:5432`, the following instrument should result:

```json
{
  "name": "db.client.duration",
  "labels": {
    "db.operation": "SELECT",

>     "db.table": "user_table",

Reviewer: Should this be
```suggestion
    "db.table": "public.user_table",
```
?
The definition above doesn't say anything about prefixes/namespaces/schemes.

Author: Excellent point.  I've included this in the trace PR.

Author: Added in cf83046.

| `net.peer.name`        | Remote hostname or similar, see note below. | `example.com` | No |

**[1]:** See the [database trace semantic conventions](../../trace/semantic_conventions/database.md#connection-level-attributes)
for the list of well-known database system values.

**[2]:** Recommended in general, required for in-process databases (`"inproc"`).

## Call-level Metric Instruments

The following metric instruments SHOULD be iterated for every database operation.

| Name                 | Instrument    | Units        | Description |
|----------------------|---------------|--------------|-------------|
| `db.client.duration` | ValueRecorder | milliseconds | The duration of the database operation. |

Database operations SHOULD include execution of queries, including DDL, DML,
DCL, and TCL SQL statements (and the corresponding operations in non-SQL
databases), as well as connect operations.

### Labels

In addition to the [common](#common) labels, the following labels SHOULD be
applied to all database call-level metric instruments.

| Label Name       | Type   | Description  | Example  | Required |
|------------------|--------|--------------|----------|----------|

> | `db.name`        | string | If no [tech-specific label](#call-level-labels-for-specific-technologies) is defined, this attribute is used to report the name of the database being accessed. For commands that switch the database, this should be set to the target database (even if the command fails). [1] | `customers`<br>`main` | Required if applicable. |

Reviewer 3: Do we need to namespace labels in general, since they already exist within the scope of the instrument? I could see an argument for them in the case of high level labels reused across different instruments. I did not namespaces in labels for the system metrics. Curious what others think about this

Author: I'd been in favor of redundancy in the label names, especially as not all the labels really apply to the database itself. 
I feel that `db.name` is more clearly the name of the database than `name` would be, I like the namespacing when these labels are siblings with things like `net.peer.name`, and I like the symmetry we get with the tracing semantics.

Author: Also, I was recently reminded of [this bit](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/common/common.md#recommendations-for-opentelemetry-authors):

> Semantic conventions exist for four areas: for Resource, Span and Log attribute names as well as Metric label keys. In addition, for spans we have two more areas: Event and Link attribute names. Identical namespaces or names in all these areas MUST have identical meanings. For example the http.method span attribute name denotes exactly the same concept as the http.method metric label, has the same data type and the same set of possible values (in both cases it records the value of the HTTP protocol's request method as a string).

Reviewer 3: Makes sense 👍

| `db.operation`   | string | The name of the operation being executed, e.g. the [MongoDB command name](https://docs.mongodb.com/manual/reference/command/#database-operations) such as `findAndModify`, or the SQL keyword. [4][5] | `findAndModify`<br>`HMSET`<br>`SELECT`<br>`CONNECT` | Required if applicable. |
| `exception.type` | string | The type of the exception (its fully-qualified class name, if applicable). The dynamic type of the exception should be preferred over the static type in languages that support it. | `java.sql.SQLException`<br/>`psycopg2.OperationalError` | Required if applicable. |

**[1]:** In some SQL databases, the database name to be used is called "schema name".

**[4]:** When setting this to an SQL keyword, it is not recommended to attempt any client-side parsing of `db.statement` just to get this property, but it should be set if the operation name is provided by the library being instrumented. If the SQL statement has an ambiguous operation, or performs more than one operation, this value may be omitted.

> **[5]:** To reduce cardinality, the value for `db.operation` should have parameters
> removed or substituted. The resulting value should be a low-cardinality value
> represeting the statement or operation being executed on the database. It may be

Reviewer 3: Seems like this is supposed to pertain to a `db.statement` parameter, which is missing in the table.

```suggestion
**[5]:** To reduce cardinality, the value for `db.operation` should leave in placeholders
and not include any actual parameter values. The resulting value should be low-cardinality
and represent the statement or operation being executed on the database. It may be
```

Might also be worth adding "it is not recommended to attempt any client-side parsing of `db.statement` to remove parameters", although users shouldn't be adding formatting them in anyway. The trace conventions say  "the value may be sanitized to exclude sensitive information," do you think that is mostly just about not substitution values?

Author: I like this. I've updated it to be more like these words. What do you think?

Reviewer 3: The `db.operation` label wouldn't have any placeholders I thought, since its just `SELECT`, `UPDATE`, etc. I was more thinking that advice would apply to `db.statement`, unless it was purposefully left out for cardinality reasons?

Reviewer 3: Without `db.statement`, is there enough granularity? In the MySQL example, all `SELECT` queries to `orders` would be aggregated into that same metric

```json
{
  "name": "db.client.duration",
  "labels": {
    "db.operation": "SELECT",
    "db.name": "ShopDb",
    "db.system": "mysql",
    "db.connection_string": "Server=shopdb.example.com;Database=ShopDb;Uid=billing_user;TableCache=true;UseCompression=True;MinimumPoolSize=10;MaximumPoolSize=50;",
    "db.user": "billing_user",
    "db.sql.table": "orders",
    "net.peer.name": "shopdb.example.com",
    "net.peer.ip": "192.0.2.12",
    "net.peer.port": "3306",
    "net.transport": "IP.TCP"
  }
}
```

}
```

against the database named "userdatabase" while connected to a MongoDB available
at `mongodb.example.com`, the following instrument should result:

```json
{
  "name": "db.client.duration",
  "labels": {
    "db.operation": "findAndModify",
    "db.table": "people",
    "db.name": "userdatabase",
    "db.system": "mongodb",
    "db.user": "the_user",
    "net.peer.name": "mongodb.example.com",
    "net.peer.ip": "192.0.2.14",
    "net.peer.port": "27017",
    "net.transport": "IP.TCP"
  }
}
```

## Connection Pooling Metric Instruments

The following metric instruments SHOULD be collected for database connection
pools. They SHOULD have all [common](#common) labels applied to them.

| Name                      | Instrument    | Units        | Description |
|---------------------------|---------------|--------------|-------------|
| `db.connection_pool.limit` | ValueObserver | {connections} | The total number of database connections available in the connection pool. |
| `db.connection_pool.usage` | ValueObserver | {connections} | The number of database connections _in use_. |

If the following detailed information is available, the following metric

> instruments MAY be collected. They SHOULD have all [common](#common) labels

Reviewer 3: ```suggestion
instruments MAY be used. They SHOULD have all [common](#common) labels
```

applied to them.

| Name                      | Instrument | Units         | Description |
|---------------------------|------------|---------------|-------------|

> | `db.connections.taken`    | Counter    | {connections} | Incremented when a connection is requested and returned from the pool.  |

Reviewer 4: These remind me of the [MemStats](https://golang.org/pkg/runtime/#MemStats) variables, which are difficult to reason about. This makes me think they should be database-specific, e.g., `db.redis.connections.taken`, maybe?

Author: OK, I think I'm on board. As I said before, I don't have strong use cases for these metrics, aside from the fact that go-redis is already collecting them. 

My only hesitation is: if other database libraries decide to pick up these metrics, and they become somewhat common, I wonder if it diminishes their value to have them namespaced separately. 
As a user, If I have redis and postgres connections, will I be frustrated that they report separately as `db.redis.connections.taken` and `db.postgres.connections.taken`?
Now that I've asked that, though, it doesn't seem very useful to aggregate connection pool information across multiple databases, so maybe it's fine to separate them.

The other option (that I'd be definitely in favor of) is to just drop these metrics entirely, so we'd only specify `db.connection_pool.limit` and `db.connection_pool.usage`. 

Which of those would you prefer?

Author: From the metrics SIG today: we're going to give guidance to use `db.{db.system}.*` for technology-specific instrument names.

And I'll include some examples using the names from go-redis.
